### ISSUE TITLE
Unit Test Coverage for Github Methods

### ISSUE THREAD

**Issue by NeonDaniel** [OpenVoiceOS/ovos_skill_manager#34]
Github URL and API parsing should have individual unit tests. Additionally, the methods for falling back from API to URL parsing should probably be tested, as well as failure cases when URL parsing fails due to rate limiting (maybe just mock tests?)

Open for input to test methodology, but rough outline I have in mind:

- Test API with token set (shouldn't hit rate limits that way)
- Test URL (with or without token? Don't see in docs if token is referenced)
- Test URL with mock returns that match rate limited warning page
- Test fallback methods without token set (might be problematic if the test runner IP gets flagged for abuse, but should result in falling back to URL methods)

**Comment by ChanceNCounter** [comment]
I think we should contact GitHub support and ask politely if the rate limiting can be disabled (or just massively increased) for a specific repo's GitHub Actions only, or for a specific token, or something. I doubt it can, but it's worth asking, because it would render that element moot.

For the rest, I agree, except that each layer of testing we add to address the rate limiting problem will push each test run closer to the rate limit. It's a catch-22.

I think mocking part of the suite is the answer. Test each potential API call exactly once, to verify we get all (*all*) the data used by the other cases. Then mock all the other cases to use data cached from the API tests. In the event the API tests fail, being GitHub actions, it should abandon the test run outright.

**Comment by NeonDaniel** [comment]
Yeah, could consider calling it a bug that their own hosted test runners are being rate limited.. Presuming they don't do anything special to generate unique IP's per-run, we could hit a rate limit because someone else was doing things (and we could have broken other actions yesterday with our repeated testing).

Per the docs, we should get 1,000 API calls per hour from this repo's actions if I understand correctly (https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting). I think it's safe to say we should be on the order of tens of calls per test run, meaning we should be able to run tens of runs per hour safely (not including any run outside of Github actions).

I agree with mocking responses; question is how much can we mock without missing some breaking change. I suppose in general this means trusting the stability of the Github API (not saying we shouldn't, just saying it's implied if we mock the returns).
